### ISSUE TITLE
``disallowed-name`` check no longer checks non-constants

### ISSUE THREAD

**Issue by mattclay** [PyCQA/pylint#3701]
Commit https://github.com/PyCQA/pylint/commit/3422e4adc718c60a467d66ab0dedf4a7032c4fc5  changed the behavior of the `invalid-name` check, but also changed the behavior of `blacklisted-name`, which appears to be unintentional (and undocumented).

### Steps to reproduce

demo.py:

```python
foo = {}.keys()
foo = None
```

`pylint demo.py -d all -e blacklisted-name`

### Current behavior

From pylint 2.5.0 and later:

```
demo.py:2:0: C0102: Black listed name "foo" (blacklisted-name)
```

### Expected behavior

From pylint 2.4.4:

```
demo.py:1:0: C0102: Black listed name "foo" (blacklisted-name)
demo.py:2:0: C0102: Black listed name "foo" (blacklisted-name)
```

### pylint --version output

pylint 2.5.0
astroid 2.4.2
Python 3.7.4 (default, Oct 16 2019, 10:35:50) 
[Clang 11.0.0 (clang-1100.0.33.8)]

**Comment by relrod** [comment]
@PCManticore I'm willing to look at this issue, as it's something that's blocking us in a number of ways. Would you be willing to set me on the right path to get started here?

It looks like the function that was modified is used by multiple of the lint checks. So off the bat I can think of two possible ways to solve it:

* Have `visit_assignname()` check and see if the check that is being run is `blacklisted-name` and adjust its logic accordingly to still emit like it used to if it is. I don't know if the function has any way to get at the currently-running check name or not, though, and I don't know if hardcoding the check name like that in the function body/logic is a feasible approach.
* Create a new `visit_assignname_const_blacklisted_name()` and have it only be used for `blacklisted-name`, and copy the chain of `if`s up to the "const" check, reverting the final one to what it was before 3422e4a.

There might be a better option than both of these, I'm not familiar enough with the pylint codebase yet. If you can point me in the right direction, I'd be able to try to tackle this regression and send a patch!

**Comment by relrod** [comment]
/cc @Pierre-Sassoulas @hippo91

Maybe one of you could point me in the right direction here? We're still hitting this and it's still blocking us on some things, any help is much appreciated :)

**Comment by Pierre-Sassoulas** [comment]
Hello @relrod, thank you for being willing to contribute, much appreciated !

The way to fix this would be to first add:
```
foo = {}.keys()  # [blacklisted-name]
foo = None  # [blacklisted-name]
```
in [blacklisted_name.py](https://github.com/PyCQA/pylint/blob/master/tests/functional/b/blacklisted_name.py)

Then regarding the implementation itself maybe moving the blacklist check to it's own checker would be more reasonable than adding another condition or a specific case. In my opinion `blacklisted-name` is a very simple check that should happen on all name so it does not **have to** be mixed with other checkers (it just is right now for some reason).

I don't have a strong opinion on this, I would merge any reasonable looking change that add the functional tests you gave as an example while still passing the old tests.

**Comment by relrod** [comment]
@Pierre-Sassoulas I took a different approach, just splitting out `blacklisted-name` and `non-ascii-name` (also affected) into helper functions and calling those when we can't call the whole `_check_name()` helper. It seemed like a way less invasive change.

CI is currently broken (independently of this change) due to [this change](https://github.com/PyCQA/astroid/pull/884) in `astroid`. I didn't attempt to debug it, mainly because I'm not sure if this is an `astroid` regression or that something needs to be updated in `pylint` to account for the breaking change.

**Comment by hippo91** [comment]
@relrod you are right concerning the CI and astroid. I will fix it ASAP.

**Comment by Pierre-Sassoulas** [comment]
Hey just to be clear, what is the exact code that is causing a problem here ? pylint still signals variable that are not in the same scope. See example here:
```python
foo = {}.keys() 
foo = None  # [blacklisted-name] 


def good_name():
    foo = "42"  # [blacklisted-name]
    return foo
```
I think having blacklisted name signaled only in one place per scope is actually better because this is less noise. Also the code in the example is problematic as 'foo' is re-declared and not used so `foo = {}.keys() ` is useless.

**Comment by relrod** [comment]
@Pierre-Sassoulas The example is a contrived one - you can reproduce it with just the `foo = {}.keys()` line which doesn't raise any error and used to:

```
(venv) rick@beryl ~/dev/ansible/pylint/pylint $ cat test.py
foo = {}.keys()
(venv) rick@beryl ~/dev/ansible/pylint/pylint $ bin/pylint test.py

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

**Comment by clavedeluna** [comment]
I think I have a fix for this, but can I get 👁️ on these test cases and lmk if they're correct?

assuming config `bad-names=foo,bar,baz,`
these would be for disallowed name

```
def baz():  # [disallowed-name]
    pass

class foo():  # [disallowed-name]
    pass

foo = {}.keys()  # [disallowed-name]
foo = 42  # [disallowed-name]

baz = []  # [disallowed-name]
bar = (1, 2,)  # [disallowed-name]
foo = Exception()  # [disallowed-name]
```

Since at this point `invalid-name` is also checked within name checker, the code change will likely change invalid name, so verifying these test cases too which would now be raised and weren't previously

```
aaa = 42  # [invalid-name]
a = lambda x: x  # [unnecessary-lambda-assignment, invalid-name
DICT = {1: lambda x: x, 2: lambda x: x + 1}]. #[invalid-name]
LAMBDA_WITH_YIELD = lambda: (yield)  # [invalid-name]
```

**Comment by Pierre-Sassoulas** [comment]
I think these functional tests make sense.

**Comment by mattclay** [comment]
The test cases look good to me as well.

**Comment by clavedeluna** [comment]
@DanielNoord Not able to access the link you provided, but I think your input comes across just fine. It seems to me this issue is then worth closing as won't fix?

**Comment by mattclay** [comment]
@DanielNoord Could you elaborate on how the working definition of a "module constant" relates to this issue?

**Comment by mattclay** [comment]
@DanielNoord Sorry, I guess I'm missing something here. Why does the definition of a "module constant" affect what the `disallowed-name` check applies to? Is it not intended to apply to things pylint considers constants?

**Comment by DanielNoord** [comment]
```python
MY_VAR = 1
my_other_var = 2  # [invalid-name]
```

The issue is that people have grown used to `MY_VAR` being allowed while `my_other_var` being disallowed.

@clavedeluna has figured out that we can't fix this issue without also disallowing `MY_VAR` using the default configuration. This would create very noticeable churn during upgrades for all users of pylint as they would need to update their configs or fix all their assignments.

**Comment by clavedeluna** [comment]
> @clavedeluna has figured out t

🤣 I'm glad that PR is useful for something because it was annoying to do!

**Comment by mattclay** [comment]
@DanielNoord @clavedeluna  I think I understand the issue with `invalid-name`, but why does that prevent fixing the issue with `disallowed-name`?

**Comment by clavedeluna** [comment]
> @DanielNoord @clavedeluna I think I understand the issue with `invalid-name`, but why does that prevent fixing the issue with `disallowed-name`?

Really good point. The sad answer is that they're both completely coupled in the name checker. While separating them out would seem to be the obvious decision, there is a good reason why they're coupled in the first place. Maybe there is some magic way to keep the behavior in https://github.com/PyCQA/pylint/pull/7808 in `_check_name` to only raise for disallowed and not invalid-name...
